src/hash_to_curve/expand_msg.rs

```diff
@@ -0,0 +1,785 @@
+            Self::Raw(buf) => buf.len(),
+        }
+    }
+}
+
+/// A trait for message expansion methods supported by hash-to-curve
+pub trait ExpandMessage: for<'x> InitExpandMessage<'x> {
+    // This intermediate is likely only necessary until GATs allow
+    // associated types with lifetimes
+}
+
+/// Trait for constructing a new message expander
+pub trait InitExpandMessage<'x> {
+    /// The state object used during message expansion
+    type Expander: ExpandMessageState<'x>;
+
+    /// Initialize a message expander
+    fn init_expand(message: &[u8], dst: &'x [u8], len_in_bytes: usize) -> Self::Expander;
+}
+
+// Automatically derive trait
+impl<X: for<'x> InitExpandMessage<'x>> ExpandMessage for X {}
+
+/// Trait for types implementing expand_message interface for hash_to_field
+pub trait ExpandMessageState<'x> {
+    /// Read bytes from the generated output
+    fn read_into(&mut self, output: &mut [u8]) -> usize;
+
+    /// Retrieve the number of bytes remaining in the generator
+    fn remain(&self) -> usize;
+
+    #[cfg(feature = "alloc")]
+    /// Construct a Vec containing the remaining bytes of the output
+    fn into_vec(mut self) -> Vec<u8>
+    where
+        Self: Sized,
+    {
+        let mut result = alloc::vec![0u8; self.remain()];
+        self.read_into(&mut result[..]);
+        result
+    }
+}
+
+/// A generator for the output of expand_message_xof for a given
+/// extendable hash function, message, DST, and output length
+pub struct ExpandMsgXof<H: ExtendableOutputDirty> {
+    hash: <H as ExtendableOutputDirty>::Reader,
+    remain: usize,
+}
+
+impl<H: ExtendableOutputDirty> Debug for ExpandMsgXof<H> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.debug_struct("ExpandMsgXof")
+            .field("remain", &self.remain)
+            .finish()
+    }
+}
+
+impl<'x, H> ExpandMessageState<'x> for ExpandMsgXof<H>
+where
+    H: ExtendableOutputDirty,
+{
+    fn read_into(&mut self, output: &mut [u8]) -> usize {
+        let len = self.remain.min(output.len());
+        self.hash.read(&mut output[..len]);
+        self.remain -= len;
+        len
+    }
+
+    fn remain(&self) -> usize {
+        self.remain
+    }
+}
+
+impl<'x, H> InitExpandMessage<'x> for ExpandMsgXof<H>
+where
+    H: Default + Update + ExtendableOutputDirty,
+{
+    type Expander = Self;
+
+    fn init_expand(message: &[u8], dst: &[u8], len_in_bytes: usize) -> Self {
+        let dst = ExpandMsgDst::<U32>::process_xof::<H>(dst);
+        let hash = H::default()
+            .chain(message)
+            .chain((len_in_bytes as u16).to_be_bytes())
+            .chain(dst.data())
+            .chain([dst.len() as u8])
+            .finalize_xof_dirty();
+        Self {
+            hash,
+            remain: len_in_bytes,
+        }
+    }
+}
+
+/// A placeholder type used to construct an ExpandMsgXmdState
+/// instance when hashing to a curve.
+#[derive(Debug)]
+pub struct ExpandMsgXmd<H: Digest>(PhantomData<H>);
```

Comment:
It comes from that fact that `ExpandMsgDst` has a lifetime so that it can reference the DST instead of copying it in, which would increase the size to ~255 bytes. This means that `ExpandMsgXmdState` has a lifetime because it potentially needs to input the DST whenever `read_into` is called (XOF only inputs it once). This seems like the best way to keep the implementations generic over an expander type without the lifetime bleeding into everything.